src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs

```diff
@@ -0,0 +1,709 @@
+            try
+            {
+                Assert.Equal(InteropHelper.GetCurrentUserName(), TestAccountUser2.AccountName, ignoreCase: true);
+            }
+            finally
+            {
+                user2ctx.Undo();
+            }
+
+            Assert.Equal(InteropHelper.GetCurrentUserName(), TestAccountUser1.AccountName, ignoreCase: true);
+        }
+        finally
+        {
+            user1ctx.Undo();
+        }
+
+        Assert.Equal(previous, InteropHelper.GetCurrentUserName());
+
+        // test with Dispose pattern
+        using (WindowsImpersonationContext ctxUser1 = CtxFactoryUser1())
+        {
+            Assert.Equal(InteropHelper.GetCurrentUserName(), TestAccountUser1.AccountName, ignoreCase: true);
+
+            (WindowsTestAccount TestAccountUser2, Func<WindowsImpersonationContext> CtxFactoryUser2) = user2Data();
+            using (WindowsImpersonationContext ctxUser2 = CtxFactoryUser2())
+            {
+                Assert.Equal(InteropHelper.GetCurrentUserName(), TestAccountUser2.AccountName, ignoreCase: true);
+            }
+
+            Assert.Equal(InteropHelper.GetCurrentUserName(), TestAccountUser1.AccountName, ignoreCase: true);
+        }
+
+        Assert.Equal(previous, InteropHelper.GetCurrentUserName());
+    }
+
+    // Impersonate doesn't behave like RunImpersonated, 
+    // RunImpersonated reverts ExecutionContext
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
+    // On full framework 'RunImpersonate' doesn't capture/revert Execution context
+    [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+    public void Impersonate_ExecutionContext_NotReverted()
+    {
+        AsyncLocal<string> impersonatedContextValue = new AsyncLocal<string>
+        {
+            Value = ""
+        };
+
+        using (WindowsIdentity.Impersonate(_fixture.TestAccount1.AccountTokenHandle.DangerousGetHandle()))
+        {
+            impersonatedContextValue.Value = "NewValue";
+        }
+
+        Assert.Equal("NewValue", impersonatedContextValue.Value);
+
+        impersonatedContextValue.Value = "";
+
+        WindowsIdentity.RunImpersonated(_fixture.TestAccount1.AccountTokenHandle, () => impersonatedContextValue.Value = "NewValue");
+
+        Assert.Equal("", impersonatedContextValue.Value);
+        Assert.NotEqual("NewValue", impersonatedContextValue.Value);
+    }
+
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
+    public async Task Impersonate_FlowExecutionContext_RunImpersonate()
+    {
+        // To be sure to not use ThreadPool thread
+        CustomTaskScheduler customTaskScheduler = new CustomTaskScheduler();
+        TaskFactory customFactory = new TaskFactory(customTaskScheduler);
+        List<bool> checkResult = new List<bool>();
+        string current = null;
+
+        await customFactory.StartNew(async () =>
```

Comment:
This isn't behaving how you're hoping it will.  You need to add `.Unwrap()` at the end of the `StartNew(…)`.  The reason for that is StartNew doesn't have a `Func<Task>`-returning overload, only a `Func<T>`, so this is going to result in StartNew returning a `Task<Task>` rather than a `Task`, meaning you're not actually `await`'ing the inner Task but rather the outer one, such that this whole operation is going to complete the moment the async lambda first yields.  By adding `.Unwrap()`, you'll take that `Task<Task>` and "unwrap" it, returning a `Task` that represents the inner one, which will then be awaited.  See https://blogs.msdn.microsoft.com/pfxteam/2012/02/08/potential-pitfalls-to-avoid-when-passing-around-async-lambdas/.

Reply:
right thank's!

Reply:
I read that blog in past...but my memory made garbage :smile: